akka-pki/src/main/scala/akka/pki/pem/DERPrivateKeyLoader.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.pki.pem
+
+import java.math.BigInteger
+import java.security.KeyFactory
+import java.security.PrivateKey
+import java.security.spec.PKCS8EncodedKeySpec
+import java.security.spec.RSAMultiPrimePrivateCrtKeySpec
+import java.security.spec.RSAOtherPrimeInfo
+import java.security.spec.RSAPrivateCrtKeySpec
+
+import akka.annotation.ApiMayChange
+import akka.pki.pem.PEMDecoder.DERData
+import com.hierynomus.asn1.ASN1InputStream
+import com.hierynomus.asn1.encodingrules.der.DERDecoder
+import com.hierynomus.asn1.types.constructed.ASN1Sequence
+import com.hierynomus.asn1.types.primitive.ASN1Integer
+
+final class PEMLoadingException(message: String, cause: Throwable) extends RuntimeException(message, cause) {
+  def this(msg: String) = this(msg, null)
+}
+
+object DERPrivateKeyLoader {
+
+  /**
+   * Converts the DER payload in [[PEMDecoder.DERData]] into a [[java.security.PrivateKey]]. The received DER
+   * data must be a valid PKCS#1 (identified in PEM as "RSA PRIVATE KEY") or non-ecrypted PKCS#8 (identified
+   * in PEM as "PRIVATE KEY").
+   * @throws PEMLoadingException when the `derData` is for an unsupported format
+   */
+  @ApiMayChange
+  @throws[PEMLoadingException]("when the `derData` is for an unsupported format")
+  def load(derData: DERData): PrivateKey = {
+    derData.label match {
+      case "RSA PRIVATE KEY" =>
+        loadPkcs1PrivateKey(derData.bytes)
+      case "PRIVATE KEY" =>
+        loadPkcs8PrivateKey(derData.bytes)
+      case unknown =>
+        throw new PEMLoadingException(s"Don't know how to read a private key from PEM data with label [$unknown]")
+    }
+  }
+
+  private def loadPkcs1PrivateKey(bytes: Array[Byte]) = {
+    val derInputStream = new ASN1InputStream(new DERDecoder, bytes)
+    // Here's the specification: https://tools.ietf.org/html/rfc3447#appendix-A.1.2
+    val sequence = {
+      try {
+        derInputStream.readObject[ASN1Sequence]()
```

Comment:
Would it make sense to nest possible exceptions in a new `PEMLoadingException`?

Reply:
I thought about it a bit but the `PEMLoadingException` will only be thrown if an unsupported content (e.g. `CERTIFICATE` or `ENCRYPTED PRIVATE KEY`) is loaded. Also when converting from PEM to DER. But once we have the `Array[Byte]` with DER only other exceptions (e.g.`IllegalArgument` or anything thrown by underlying libraries) is thrown.